### ISSUE TITLE
Very low wps with H200 Gpus

### ISSUE THREAD

**Issue by aniltrkkn** [pytorch/torchtitan#676]
Hello, I am running the multinode_trainer.slurm (llama3_70b.toml) on 4 nodes that have 32  H200 Gpus. However, wps is only around ~200. Any ideas what can cause this slowness?

[output.txt](https://github.com/user-attachments/files/17740634/output.txt)

[multinode_trainer.slurm.txt](https://github.com/user-attachments/files/17740601/multinode_trainer.slurm.txt)

**Comment by lessw2020** [comment]
Hi @aniltrkkn  - hard to say without a trace, but most likely something is amiss on the 'between node' connections for your setup and creating this slowness. 

You mentioned you are using the multinode_slurm script - are you running this on AWS?  There are some settings in there that were to ensure EFA is used for cross node comms, but it was never tested with H200 as AWS did not have them at that point and we no longer have AWS cluster access.   

AWS has up to 3200 GBs for H100/H200 but it could be the settings need to be adjusted for EFA with the new H200s.  

A couple options here:
a - if you can confirm you are on AWS (please confirm exactly what you are using ala EC2,  etc)  I can reach out to their SA's to review the multi-node slurm script and see what might need adjusting.

b - if you are not on AWS, you could adjust it directly for your hardware as the script assumes EFA is available so it might need tuning to leverage your higher speed node interconnect.  Assuming my guess re: the likel issue is the between node network speed is correct. 

c - finally, you could also run the same test you ran above but turn on profiling in the toml and get a trace or two and that would confirm where the slowdown is. You can gz compress a trace and post here as it should shrink it down to minor size and happy to take a look.

**Comment by lessw2020** [comment]
btw, a quick test would also be just run the same short run on llama3-8b with FSDP only and using a single node and lets see how your wps looks there.  That should be quite fast but if that is also slow, then the issue is within node rather than between node and would help ensure we bisecting the issue properly.

**Comment by aniltrkkn** [comment]
Hi @lessw2020, Thank you very much for your response. Here are my responses to your questions:

I am getting high wps with single node 8B trainings.

We are not using AWS for trainings so I need to check if we EFA is available in our training datacenter. But, our multi node training code works fine in the same cluster. 

I am attaching the profile trace for one of the multi-node 70B trainings.
[profile_trace.tar.gz](https://github.com/user-attachments/files/17756342/profile_trace.tar.gz)

This may be not related to the issue but we are training 70B with a mesh of ['dp', 'tp'], [4, 8] (4 nodes). I always thought each data parallel group will have its own shards. But when I increase the number of training nodes to 5, I get this error:

"FSDP+TP sharding does not support uneven sharding for now: tensor dim 0 has size 8192 which cannot be evenly sharded into 40 shards"

It seems like we are sharding the model into 40 instead of 5. Is this the intended behavior?

**Comment by yifuwang** [comment]
Hmm the slurm script you posted says CUDA_LAUNCH_BLOCKING=0, but the trace looked like it was run with CUDA_LAUNCH_BLOCKING=1. Could you double check this?

**Comment by aniltrkkn** [comment]
Hi @yifuwang, i made it 0 and it is still very low. It seems like we don't support EFA, and we use InfiniBand. I tried our regular parameters 

export CUDA_LAUNCH_BLOCKING=0
export NCCL_SOCKET_IFNAME=eth0
export NCCL_IB_HCA=ibp
export NCCL_MIN_CTAS=32
export UCX_NET_DEVICES=ibp0:1,ibp1:1,ibp2:1,ibp3:1,ibp4:1,ibp5:1,ibp6:1,ibp7:1
export SHARP_COLL_ENABLE_PCI_RELAXED_ORDERING=1
export NCCL_COLLNET_ENABLE=0
export NCCL_DEBUG=INFO
export NCCL_ALGO=NVLSTREE

but they also do not help. Is EFA absolutely necessary?

**Comment by awgu** [comment]
@aniltrkkn Could you share the new trace? Either way, you are heavily communication bound (e.g. FSDP all-gather is >2x longer than the forward compute). Perhaps, could you try HSDP (no TP) with 8-way sharding to keep the FSDP all-gather/reduce-scatter within node?

**Comment by casper-hansen** [comment]
@aniltrkkn I struggled with the same issue, but fixed it by specifying the right environment variables and installing a library

Related:
https://github.com/pytorch/torchtitan/issues/708#issuecomment-2506388356

**Comment by lessw2020** [comment]
fyi, I'm running on H200's atm with very good tps.  have run up to 1600 with HSDP2 in fact.  
Thus this is a config specific issue rather than an FSDP2/HSDP2 issue.  
A starting recommendation is to run the NCCL test and confirm your bandwidth in your cluster. 
I'm going to close out for now since this not H200 specific but cluster and config specific.
